### PR TITLE
Cleanup #1982: Use C++ STL data structures in conv-conds

### PR DESCRIPTION
**Pull request by evan-charmworks**
*Original date: 2019-04-26 04:00:57*
*Original PR: https://charm.cs.illinois.edu/gerrit/4932*

---

Change-Id: Ie533057d1dba8db277388affce245dc8d2cb50e6

### CODE REVIEW

**Comment by stwhite91** [comment]
Performance seems to be equivalent to the original C implementation in microbenchmarks, but this PR is failing tests/charm/megatest with:

"------------- Processor 0 Exiting: Called CmiAbort ------------
Reason: Internal QD Error. Contact Developers.!"

**Comment by stwhite91** [comment]
Fixed, it passes all tests now.

**Review by stwhite91 — APPROVED** [review]

**Comment by stwhite91** [comment]
Fine-grained performance measured using Charm 1D array pingpong on 1 PE is unchanged (within the noise) with this PR from master.

**Review by stwhite91 — APPROVED** [review]

**Comment by evan-charmworks** [comment]
Strange segfault in the Spack build. Unfortunately it doesn't show the most important part of the backtrace.

```
     22403      [1:15] bcast_subset_async_delayed 0x5555557717fd CsdScheduleForever
     22404      [1:16] bcast_subset_async_delayed 0x555555771a05 CsdScheduler
     22405      [1:17] bcast_subset_async_delayed 0x5555557bbdda
     22406      [1:18] bcast_subset_async_delayed 0x5555557bbe42
     22407      [1:19] libpthread.so.0 0x7ffff7f9d609
     22408      [1:20] libc.so.6 0x7ffff7b52133 clone
  >> 22409    Fatal error on PE 1> unknown signal
     22410    
     22411    real	0m0.230s
     22412    user	0m0.004s
     22413    sys	0m0.000s
  >> 22414    make[5]: *** [Makefile:48: smptest] Error 1
     22415    make[5]: Leaving directory '/tmp/runner/spack-stage/spack-stage-c
              harmpp-main-gqo6lhvz3f3e4gvegepesfz2szln437h/spack-src/tests/char
              m++/zerocopy/zc_post_async/bcast_subset_async'
```

**Comment by stwhite91** [comment]
Looking at that ZC Post API Bcast test, it doesn't use any CcdCallbacks on its own (only what internal callbacks are used by the runtime).

File: src/conv-core/conv-conds.C [hunk 1]
```diff
@@ -401,26 +247,23 @@ static void ccd_heap_remove(void)
  */
 static void ccd_heap_update(double curWallTime)
 {
-  ccd_heap_elem *h = CpvAccess(ccd_heap);
-  ccd_heap_elem *e = h+CpvAccess(ccd_heapmaxlen);
-  int i,ne=0;
+  auto & h = CpvAccess(ccd_heap);
+
+  std::vector<ccd_periodic_callback> expired;
```
Comment (stwhite91) [inline src/conv-core/conv-conds.C:252]: Thinking about this more, we might want to hoist this vector to persistent state in order to avoid small heap allocations everytime we pull something out of the heap. Could be in a separate commit.

Comment (evan-charmworks) [reply src/conv-core/conv-conds.C]: Despite the comment in this function, I don't see why we need to process the heap elements in two phases if handled correctly. Let's see if CI validates my thinking.

**Comment by evan-charmworks** [comment]
The issue was actually a bug in my reimplementation. By default, `std::priority_queue` sorts in descending order, whereas we want to sort the time field in ascending order. I fixed that in the main commit. I also pushed other minor cleanup I made while debugging.

**Review by stwhite91 — APPROVED** [review]

**Review by matthiasdiener — APPROVED** [review]
